### PR TITLE
sql,*: rework plan hooks to type check separately from execution

### LINKED ISSUE #88385: Panic when running UNION on SHOW BACKUP queries as non-root user
**Describe the problem**

If a query issued by a non-root user UNIONs the result of multiple SHOW BACKUP queries, then CRDB crashes with an exit code of 2.

**To Reproduce**

1. Create a user and grant perms.
```
root@localhost:26257/defaultdb> create user test;
CREATE ROLE


Time: 25ms total (execution 25ms / network 0ms)

root@localhost:26257/defaultdb> grant admin to test;
GRANT


Time: 29ms total (execution 29ms / network 0ms)
```
2. Take a backup
```
test@localhost:26257/defaultdb> BACKUP INTO 'userfile:///backups';
        job_id       |  status   | fraction_completed | rows | index_entries | bytes
---------------------+-----------+--------------------+------+---------------+--------
  798619757264404481 | succeeded |                  1 |   24 |             9 |  1636
(1 row)


Time: 466ms total (execution 466ms / network 0ms)

test@localhost:26257/defaultdb> SHOW BACKUPS in 'userfile:///backups';
          path
-------------------------
  /2022/09/21-195008.34
(1 row)


Time: 23ms total (execution 23ms / network 0ms)
```
3. Union the result of 2 SHOW BACKUP queries, this might take a few tries to get the race condition to occur.
```
test@localhost:26257/defaultdb> SELECT * FROM [SHOW BACKUP LATEST IN 'userfile:///backups'] UNION SELECT * FROM [SHOW BACKUP LATEST IN 'userfile:///backups'];
ERROR: unexpected EOF
warning: error retrieving the transaction status: connection closed unexpectedly: conn closed
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the database name: failed to connect to `host=localhost user=test database=`: dial error (dial tcp [::1]:26257: connect: connection refused)
```
4. CRDB should have crashed. The interesting stack traces should be the ones that include the backup call.

```
I220921 13:50:12.719035 1 util/log/flags.go:211  [-] 1  stderr capture started
fatal error: concurrent map read and map write

goroutine 11141 [running]:
github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv.(*StoredCatalog).UpdateValidationLevel(0xc005dc0580, {0x7fd447d2a088, 0xc0000d5ef0}, 0x7)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catkv/stored_catalog.go:374 +0x94
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).finalizeDescriptors(0xc005dc04e0, {0x5f95168, 0xc001b42720}, 0x0?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:395 +0x45b
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getByName(0xc005dc04e0, {0x5f95168, 0xc001b42720}, 0xc0028de2c0, {0x0?, 0x0?}, {0x0?, 0x0?}, {0x510253f, 0x6}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:324 +0x5c5
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getDatabaseByName(0x203000?, {0x5f95168?, 0xc001b42720?}, 0x0?, {0x510253f, 0x6}, {0x0, 0x0, 0x0, 0x0, ...})
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/database.go:64 +0x88
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetImmutableDatabaseByName(...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/database.go:45
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getObjectByNameIgnoringRequiredAndType(0xc005dc04e0, {0x5f95168, 0xc001b42720}, 0xc005868f58?, {0x510253f, 0x6}, {0x5101fc9, 0x6}, {0x511b386, 0xc}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/object.go:160 +0x13d
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getObjectByName(0x2?, {0x5f95168, 0xc001b42720}, 0x0?, {0x510253f?, 0xc45f80?}, {0x5101fc9?, 0x1945786?}, {0x511b386, 0xc}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/object.go:65 +0x1b8
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getTableByName(0x1000000000000?, {0x5f95168, 0xc001b42720}, 0x17303c0?, {0x5fc7720, 0x9809800}, {{0x1, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/table.go:53 +0x119
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetImmutableTableByName(0xc002e9ce90?, {0x5f95168?, 0xc001b42720?}, 0xc003207400?, {0x5fc7720?, 0x9809800?}, {{0x1, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/table.go:44 +0x4e
github.com/cockroachdb/cockroach/pkg/sql.MemberOfWithAdminOption({0x5f95168?, 0xc001b42720}, 0xc0013c0a00, {0x5fe4430?, 0xc0005a33b0}, 0xc0058690a8?, 0xc0028de2c0, {{0xc00180c040?, 0x0?}})
	github.com/cockroachdb/cockroach/pkg/sql/authorization.go:462 +0x14c
github.com/cockroachdb/cockroach/pkg/sql.(*planner).MemberOfWithAdminOption(0x0?, {0x5f95168?, 0xc001b42720?}, {{0xc00180c040?, 0x0?}})
	github.com/cockroachdb/cockroach/pkg/sql/authorization.go:433 +0x65
github.com/cockroachdb/cockroach/pkg/sql.(*planner).UserHasAdminRole(0x0?, {0x5f95168?, 0xc001b42720?}, {{0xc00180c040?, 0x0?}})
	github.com/cockroachdb/cockroach/pkg/sql/authorization.go:393 +0x88
github.com/cockroachdb/cockroach/pkg/cloud/cloudprivilege.CheckDestinationPrivileges({0x5f95168, 0xc001b42720}, {0x6025240, 0xc001b1d0d0}, {0xc000529820, 0x1, 0x0?})
	github.com/cockroachdb/cockroach/pkg/cloud/cloudprivilege/privileges.go:29 +0x94
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.showBackupPlanHook.func1({0x5f95168, 0xc0083a2db0}, {0x0?, 0x0?, 0xf0f00c00a3cbf68?}, 0x5f95168?)
	github.com/cockroachdb/cockroach/pkg/ccl/backupccl/show.go:279 +0x30a
github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1()
	github.com/cockroachdb/cockroach/pkg/sql/planhook.go:184 +0xa4
created by github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec
	github.com/cockroachdb/cockroach/pkg/sql/planhook.go:182 +0x1b2
```

**Expected behavior**

There should be no crash!

**Additional data / screenshots**

Log directory: https://upload.cockroachlabs.com/receive/?packageCode=M1KruQE98L65Cm1oy6GPshIXN29XnAdZeaRNvTreTgw#keycode=GgD909dEchER8yB1YbmTUiksTlrL24aaRknToo7MrXk

**Environment:**
```
Build Tag:        v22.2.0-alpha.3-265-gedeefa9120
Build Time:       2022/09/20 07:10:37
Distribution:     CCL
Platform:         linux amd64 (x86_64-pc-linux-gnu)
Go Version:       go1.19.1
C Compiler:       gcc 6.5.0
Build Commit ID:  edeefa9120e86739808ca43fc52d4cbac813fab3
Build Type:       release
```

**Additional context**

This was discovered while running some performance tests on backup queries.

### PR DESCRIPTION
**Pull request by ajwerner**
Prior to this change, plan hooks would be invoked twice, once during prepare and again during execution. At prepare time, the placeholder values are not known, but, critically, the placeholders were being type-checked. This type checking informs the sql layer how to interpret the placeholder values when we are trying to bind them. Importantly, the sql layer does not ever use the `planNode` produced via the `PlanHookRowFn` returned from the planning invocation.

The above dance meant that these plan hooks had quite a bit of complexity to eagerly type check but lazily evaluate expressions passed in via the statement. This complexity became a problem more recently: one cannot and should not access many planner data structures from the row func because such functions can, theoretically, run concurrently and these data structures are not thread safe.

Note that the above problem is not commonly a real problem. Most plan hook statements don't permit concurrency and those than do, via, say, `UNION ALL` don't commonly get executed in such a setting. In some ways, the real reason for this change is to unwind some unneeded complexity and clarify the bounds of what one can do in a plan hook.

The big change which unearthed the problem is that we now support privileges on external URLs. These URLs may be provided via placeholders or expressions. If we couldn't access the expressions until we were in the row fn, then we'd be at risk of performing privilege checking concurrently, which is not safe.

One thing that this commit does is introduce a new package `exprutil` with helpers for type checking. It is so named because I'd like to also remove the methods from the `PlanHookState` which we now use for expr evaluation and place them in this new package. 

Fixes #88385

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
I will rename `evalexpr` to `exprutil`

**Review by dt — APPROVED** [review]
Looked though the backup changes and looks great; none of our hook plan statements are expected to be prepare-once-execute-repeatedly, so just doing all our work in exec makes sense to me, with a super-minimal type-check-only prep phase. Doing permission/auth checks at exec makes sense for hook plans since some of the permissions we check are based on the _values_ of bound parameters (e.g. the URI to which one is backing up) to overall this looks sane to me.

File: pkg/ccl/backupccl/backup_test.go [hunk 1]
```diff
@@ -8640,7 +8640,7 @@ func TestRestorePauseOnError(t *testing.T) {
 	sqlDB.Exec(t, `DROP DATABASE r1`)
 
 	restoreQuery := `RESTORE DATABASE r1 FROM 'nodelocal://0/eventlogging' WITH DEBUG_PAUSE_ON = 'error'`
-	findJobQuery := `SELECT job_id FROM [SHOW JOBS] WHERE description LIKE '%RESTORE DATABASE%' ORDER BY created DESC`
+	findJobQuery := `SELECT job_id FROM [SHOW JOBS] WHERE description LIKE '%RESTORE DATABASE%' AND job_type = 'RESTORE' ORDER BY created DESC`
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/ccl/backupccl/backup_test.go]: Fixed

**Review by ajwerner — COMMENTED** [review]
Alright, I cleaned this up. Should be RFAL.

**Review by stevendanna — APPROVED** [review]
I like this. seems like it'll also lower some of the psychological overhead of supporting placeholders for more parts of the planHook driven statements.

**Review by miretskiy — APPROVED** [review]

**Comment by ajwerner** [comment]
@rafiss and @mgartner can I get some some SQL eyes on this?

**Review by rafiss — CHANGES_REQUESTED** [review]
this looks nice! i think the new package  looks better to use

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-:-NEpvb0lBbSzX0CcESrI:b-u3en7r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @rhu713)

___
*[`pkg/sql/alter_role.go` line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpqFpa6lR8ZELhh3xT:-NEpqFpa6lR8ZELhh3xU:bgerg9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/alter_role.go#L74)):*
> ```Go
> }
> 
> func asStringOrNull(
> ```

the nested `return func` structure here makes it bit hard to follow. could you add comments, or even making the function return values be named parameters could be sufficient

also, since it seems to be only used for role options, would this make sense to put in the `sql/roleoption` package?
___
*[`pkg/sql/alter_role.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpuK4lEvi93Bt1cDBO:-NEpuK4lEvi93Bt1cDBP:b-f95psy) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/alter_role.go#L90)):*
> ```Go
> 				return true, "", nil
> 			}
> 			return false, string(*got.(*tree.DString)), nil
> ```

want to use `string(tree.MustBeDString(d))`?
___
*[`pkg/sql/exprutil/evaluator.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpsZsB-D-UsjiE2vw_:-NEpsZsB-D-UsjiE2vwa:b-8xzlng) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L35)):*
> ```Go
> 
> // StringArray evaluates expr to a string slice.
> func (e Evaluator) StringArray(ctx context.Context, exprs tree.Exprs) ([]string, error) {
> ```

want to use pointer receivers? i'm just thinking about  avoiding copying the `op` string unnecessarily, which to be fair probably doesn't matter at all
___
*[`pkg/sql/exprutil/evaluator.go` line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEptXKh0KJpqPzkLdn9:-NEptXKh0KJpqPzkLdnA:b-f95psy) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L53)):*
> ```Go
> 		return "", err
> 	}
> 	return string(*d.(*tree.DString)), nil
> ```

want to use `string(tree.MustBeDString(d))`?
___
*[`pkg/sql/exprutil/evaluator.go` line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEptaGn7BkvG1F4fc8U:-NEptaGn7BkvG1F4fc8V:bvjtita) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L62)):*
> ```Go
> 		return false, err
> 	}
> 	return bool(*d.(*tree.DBool)), nil
> ```

want to use `bool(tree.MustBeDBool(d))`?
___
*[`pkg/sql/exprutil/evaluator.go` line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpuyqyAYfFKYJnh7qJ:-NEpuyqyAYfFKYJnh7qK:b3xlv0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L72)):*
> ```Go
> 	for _, opt := range opts {
> 		k := string(opt.Key)
> 		validate, ok := optValidate[k]
> ```

instead of the validation logic below, should this call `optValidate.validate(opt)`

**Review by ajwerner — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-:-NEqPO544cP1W_cU_Lzc:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @rafiss, and @rhu713)

___
*[`pkg/sql/alter_role.go` line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpqFpa6lR8ZELhh3xT:-NEqKQ3M3Je1Z3hLw9Ox:bk6484) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/alter_role.go#L74)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the nested `return func` structure here makes it bit hard to follow. could you add comments, or even making the function return values be named parameters could be sufficient

also, since it seems to be only used for role options, would this make sense to put in the `sql/roleoption` package?
</blockquote></details>

I unwound this mess, moved some of the logic to `exprutil` and also reversed the `roleoption`/`tree` import direction. See what you think. 
___
*[`pkg/sql/alter_role.go` line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpuK4lEvi93Bt1cDBO:-NEqKY3X4i3yK0eInddl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/alter_role.go#L90)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

want to use `string(tree.MustBeDString(d))`?
</blockquote></details>

Done.
___
*[`pkg/sql/exprutil/evaluator.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpsZsB-D-UsjiE2vw_:-NEqKYl-EiUUWoWKdSZc:bn0t7m2) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L35)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

want to use pointer receivers? i'm just thinking about  avoiding copying the `op` string unnecessarily, which to be fair probably doesn't matter at all
</blockquote></details>

Meh it means you can't use it inline. I don't feel like optimizing that right now. 
___
*[`pkg/sql/exprutil/evaluator.go` line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEptXKh0KJpqPzkLdn9:-NEqKZigCU1tGzcpBFf2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L53)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

want to use `string(tree.MustBeDString(d))`?
</blockquote></details>

Done.
___
*[`pkg/sql/exprutil/evaluator.go` line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEptaGn7BkvG1F4fc8U:-NEqK_c-1Gc6wh3zD5cs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L62)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

want to use `bool(tree.MustBeDBool(d))`?
</blockquote></details>

Done.
___
*[`pkg/sql/exprutil/evaluator.go` line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-NEpuyqyAYfFKYJnh7qJ:-NEqKsjTD124Zf-wKQJA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50e3122b66d67c7fee101102a02580d0259aaa62/pkg/sql/exprutil/evaluator.go#L72)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of the validation logic below, should this call `optValidate.validate(opt)`
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
lgtm! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90254#-:-NEqSQWW44xz3aUKk1cA:b-nzoors)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @rhu713)

**Comment by ajwerner** [comment]
TFTR!

bors r+
